pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,197 @@
+        return model_trace.log_pdf()
+
+    def grad_potential(self, z):
+        log_joint_prob = self.log_prob(z)
+        log_joint_prob.backward()
+        grad_potential = {}
+        for name, node in z.iter_stochastic_nodes():
+            grad_potential[name] = -node['value'].grad.clone()
+            grad_potential[name].volatile = False
+        return grad_potential
+
+    def _setup(self, z):
+        # store the current value of param gradients so that they
+        # can be reset at the end
+        for name, node in z.iter_param_nodes():
+            self.cached_param_grads[name] = node['value'].grad
+
+    def _cleanup(self):
+        # reset the param values to those stored before the hmc run
+        for name, grad in self.cached_param_grads.items():
+            param = pyro.get_param_store().get_param(name)
+            param.grad = grad
+
+    def energy(self, z, r):
+        kinetic_energy = 0.5 * torch.sum(torch.stack([r[name] ** 2 for name in r]))
+        potential_energy = - self.log_prob(z)
+        return kinetic_energy + potential_energy
+
+    def _traces(self, *args, **kwargs):
+        t = 0
+        self.args = args
+        self.kwargs = kwargs
+        # initialize z with model trace
+        z = poutine.trace(self.model).get_trace(*args, **kwargs)
+        self._setup(z)
+        # sample p's from the distribution given by p_dist
+        r_dist = {}
+        for name, node in z.iter_stochastic_nodes():
+            r_mu = torch_zeros_like(node['value'])
+            r_sigma = torch_ones_like(node['value'])
+            r_dist[name] = dist.Normal(mu=r_mu, sigma=r_sigma)
+        self.accept_cnt = 0
+
+        # Run HMC iterations
+        print('Simulating using HMC...')
+        while t < self.warmup_steps + self.num_samples:
+            if t % 100 == 0:
+                print('Iteration: {}'.format(t))
+            # sample momentum
+            r = {name: r_dist[name].sample() for name in z.stochastic_nodes}
+            z_new, r_new = verlet_integrator(z,
+                                             r,
+                                             self.grad_potential,
+                                             self.step_size,
+                                             self.num_steps)
+            energy_proposal = self.energy(z_new, r_new)
+            energy_current = self.energy(z, r)
+            # print("Energy - current: {}".format(energy_current))
+            # print("Energy - proposal: {}".format(energy_proposal))
+            delta_energy = energy_proposal - energy_current
+            t += 1
+            if random.random() < (-delta_energy).exp().data[0]:
+                self.accept_cnt += 1
+                z = z_new
+            if t < self.warmup_steps:
+                continue
+            yield (z, z.log_pdf())
+        self._cleanup()
+
+    def get_acceptance_ratio(self):
+        return self.accept_cnt / float(self.warmup_steps + self.num_samples)
```

Comment:
nit: if you `from __future__ import division`, then you won't need the `float` cast :wink: